### ISSUE TITLE
Support stealth address

### ISSUE THREAD

**Issue by Keith-CY** [Magickbase/neuron-public-issues#100]
Stealth Address is an interesting use case of UTXO and is a prospective feature of Neuron(and Nexus).

We can take a brief look at its definition and use cases. Then if you are interested, go deeper.

Ref: https://www.investopedia.com/terms/s/stealth-address-cryptocurrency.asp

cc @yanguoyu @JeffreyMa597 @homura @zhangyouxin @IronLu233 @zhb666 @Danie0918

**Comment by Keith-CY** [comment]
We may start stealth address mode with light client because light client has limited ability to on-chain data, while stealth address limits the usage of HD wallet.

For instance, in stealth mode, the path of address generation is predictable, so light client doesn't have to subscribe to all addresses simultaneously, and once a transaction is bound to an address, it could be marked as used and removed from the subscription.

**Comment by Keith-CY** [comment]
Some details were added at https://github.com/Magickbase/neuron-public-issues/issues/100#issuecomment-1517652586

Please check if it's feasible.

Besides, the stealth mode should also be considered in Nexus because the rule of key generation is going to be the same. @homura

**Comment by zhangyouxin** [comment]
After reading about [Monero Steal Address](https://www.getmonero.org/resources/moneropedia/stealthaddress.html) and [bitcoin steal address](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2014-January/004020.html), I think the current neuron wallet has provided some privacy with advantage of UTXO model and BIP32, because a user need `extended-pubkey` to view all incomming transactions, which is equivilant to `view key` in Monero. The key difference here is that user should use address that is not revealed on-chain to receive assets

So I have some questions for the core rule here: 

 - Is the steal mode in neuron aiming to use an address only once to receive token?
 - how to generate a steal address by sender when given a receiver public address? 
 - why is there a block count down after an address is revealed?

**Comment by homura** [comment]
An unpublished [proposal](https://github.com/genjix/bips/blob/master/bip-stealth.mediawiki) for Bitcoin

**Comment by Keith-CY** [comment]
> So I have some questions for the core rule here:
> 
> * Is the steal mode in neuron aiming to use an address only once to receive token?

Yes, each address is expected to be a one-time address, but it's hard to reject transfers to a specific address, so a countdown will be introduced to freeze the address, as mentioned in point 2 and 3 at https://github.com/Magickbase/neuron-public-issues/issues/100#issuecomment-1517652586

> * how to generate a steal address by sender when given a receiver public address?

I've considered concealing the address against the third party, but we can have a talk about it

> * why is there a block count down after an address is revealed?

It's for the light client mode, it's up to 500 scripts to subscribe with the light client so the activated addresses should be limited. With the countdown, we can keep the group of subscribed scripts small

**Comment by yanguoyu** [comment]
I think full node can also support stealth address because `the path of address generation is predictable`.
And we should keep the original address mode and add a stealth address.

**Comment by Keith-CY** [comment]
> I think full node can also support stealth address because `the path of address generation is predictable`. And we should keep the original address mode and add a stealth address.

Sure, stealth address mode is a mode of address generation, light-client/full-node is a mode of synchronization, these two modes are independent and synchronization mode is agnostic to address generation.

Stealth address mode is mentioned along with light client above because the limit of light client could be ameliorated by stealth address mode. It doesn't mean that stealth address mode is only available with light client

**Comment by Keith-CY** [comment]
Here's a private note coming from the core team, it's about implementing stealth address by key-exchange solution
ref: https://docs.google.com/document/d/18CgG_bjweOx0ZgI-eFRFn5HYWOLWh-BFfjBQ-aRfLl8/edit?usp=sharing

**Comment by Keith-CY** [comment]
> Here's a private note coming from the core team, it's about implementing stealth address by key-exchange solution ref: [docs.google.com/document/d/18CgG_bjweOx0ZgI-eFRFn5HYWOLWh-BFfjBQ-aRfLl8/edit?usp=sharing](https://docs.google.com/document/d/18CgG_bjweOx0ZgI-eFRFn5HYWOLWh-BFfjBQ-aRfLl8/edit?usp=sharing)

Here's a feasibility report about implementing stealth address in Neuron based on this document

-----------

Stealth address is a concept that consists of two concrete elements. Here we introduce two terms from https://eips.ethereum.org/EIPS/eip-5564#specification for clarity 
- **stealth meta-address**: the address published by the address owner, it includes necessary factors to generate a random lock script to dominate an asset, the random lock script will be mapped to a **stealth recipient-address**. The address base only used to share message between sender and receiver, and won’t be used as a recipient directly.
- **stealth recipient-address**: the ckb address encoded from a random lock script derived from a **stealth meta-address**. It’s used to dominate an asset and should be used as a recipient directly.

# How to manage assets with stealth address

1. generate **stealth meta-address**
2. publish **stealth meta-address**
3. match **stealth recipient-address**s on chain
4. manage assets dominated by **stealth recipient-address**s
5. send to **stealth recipient-address**

## Generate **stealth meta-address**

There are 3 factors in generating a **stealth meta-address**
- Lock script to specify the rule of asset domination
- Private key(named *q* in the doc) to dominate the address recipient
- Viewing key(named *v* in the doc) to match on-chain **stealth recipient-address**s

The first 2 factors are clear.

Lock script is determined by the business logic, and it can be any one that conforms to stealth address lock script

Private key *q* dominates the address, so it should be one of the private keys derived from mnemonics.

Viewing private key *v* is a bit special because it should be public in the runtime so password can be avoided during sign-and-match stage of scanning on-chain **stealth recipient-address**s

1. *v* should not be one of the derived private keys because those keys may be derived in another wallet and used to dominate assets. If *v* is public in Neuron, the assets dominated by *v* will be in danger.
2. *v* should be migrated along with the keystore when the wallet is backed up. Otherwise, the **stealth meta-address** cannot be reproduced in another wallet because *v* is one factor of *c*, the shared secret.
3. *v* should not be shared between **stealth meta-address**s, otherwise, all derived **stealth recipient-address**s can be tracked by the same *v*

Ideally, *v* should have a determined relation to *q* while *q* won’t be revealed by *v*, intuitively get 

```js
const v = hash(q)
```

Here we get the lock script of a **stealth meta-address**

```yml
code_hash: lock_base.code_hash
hash_type: lock_base.hash_type
args: V | Q
```

## Publish **stealth meta-address**

// TODO

## Match **stealth recipient-address**s on chain

Once a **stealth meta-address** is generated, it can be used to scan on-chain assets with the following logic

1. check if an on-chain lock script matches pattern of a stealth address lock script, if yes
2. compute the shared secret *c* by *H(v.P)*
3. validate the lock script by tweaked private key *v + c*, if yes
4. the lock is derived from “stealth meta-address” of *v* and should be synced

## Manage assets dominated by **stealth recipient-address**s

Similar to “sudt account” in Neuron, every “stealth meta-address” could be considered an independent sub-account that consists of infinite “stealth recipient-address”s, and their balances should be grouped by “stealth meta-address”

## Send to “stealth recipient-address”

When an inputted address is recognized as a “stealth meta-address”, it should be transformed into a “stealth meta-lock-script” as follows

```yml
code_hash: stealth meta-address.code_hash
hash_type: stealth meta-address.hash_type
args: Q | V
```

and derives a “stealth recipient-lock-script”  with a random private key ‘p’ as follows

```yml
code_hash: stealth meta-address.code_hash
hash_type: stealth meta-address.hash_type
args: P | Q’ | V’ # P = publicKey(p), Q’ = Q + p.Q.G, V’ = V + p.V.G
```

By that the **stealth recipient-address** is determined

**Comment by Danie0918** [comment]
# Neuron Product Planning

## Design Principles

1. Ease of use: Considering the advanced needs of stealth address as a user's asset management, making it an advanced function under one account will be more convenient for users to manage their assets while facilitating subsequent expansion.
2. Multiplicity: the need exists for users to manage different assets with multiple stealth addresses, thus a single wallet account can create multiple privacy addresses.
3. Help users avoid errors: need to consider user account and device change scenarios, need to add guidance when users add new stealth address accounts, export wallets.

## Detailed planning

1. **Send page** - Add stealth transaction mode

Send page to add stealth transaction mode, the user needs to enter the stealth meta-address of the receiving asset and synthesize the address of the receiving asset according to the random number automatically generated by the software, and enter the asset to send.

2. **Add Stealth Address Account** - Users can create a Stealth Address account and receive and manage their assets under this page.

- Create stealth address accounts : Users can create 1-N stealth address accounts, and the software automatically generates corresponding secret keys and stealth meta-addresses.
- Receiving stealth address assets : Users can scan for stealth transactions and get feedback by clicking on the receiving asset in their stealth address account.
- Managing stealth assets : Users can view the assets under the account in the stealth account and unlock and send the assets.

3. Anti-errors guidance

- Create Stealth Accounts : Guides users to back up their wallets so that they don't lose their stealth accounts.
- Export Wallet : Export the private key containing the stealth account and prompt the user for the included stealth account.

**Comment by quake** [comment]
In the original design scheme, q and v are independent, if v is derived from q: v = H(q), then it add the correlation between v and q. 

Assuming that someone gets v by attacking the wallet, it is possible to reduce the difficulty of breaking q based on my light understanding of discrete logarithms, e.g., by using [Pollard's rho algorithm](https://en.wikipedia.org/wiki/Pollard%27s_rho_ algorithm)

**Comment by Keith-CY** [comment]
> In the original design scheme, q and v are independent, if v is derived from q: v = H(q), then it add the correlation between v and q.
> 
> Assuming that someone gets v by attacking the wallet, it is possible to reduce the difficulty of breaking q based on my light understanding of discrete logarithms, e.g., by using [Pollard's rho algorithm](https://en.wikipedia.org/wiki/Pollard%27s_rho_algorithm)

I'm still trying to understand the [Pollard's rho algorithm](https://www.wikiwand.com/en/Pollard%27s_rho_algorithm), but there is a mature algorithm to protect parent keys from revealed child keys. As mentioned in section **Hardened child key derivation** at https://www.oreilly.com/library/view/mastering-bitcoin/9781491902639/ch04.html

> This creates a “firewall” in the parent/child sequence, with a chain code that cannot be used to compromise a parent or sibling private key.

We may use a hardened-derived child key of *q* as the viewing key *v*

**Comment by quake** [comment]
in bip32, we need to use parent chain code and private key to generate children, the `chain code` is used to introduce entropy to the hash function. I think it's unsafe to use `v = H(q)` directly, it's same as using `child_pk = H(parent_key)` without entropy.

**Comment by Keith-CY** [comment]
> in bip32, we need to use parent chain code and private key to generate children, the `chain code` is used to introduce entropy to the hash function. I think it's unsafe to use `v = H(q)` directly, it's same as using `child_pk = H(parent_pk)` without entropy.

Indeed, my presentation should be rectified as follows:

Ideally, *v* should have a determined relation to *q* while *q* won’t be revealed by *v, intuitively get

```js
v = H(q, e)
```

where *e* is an entropy to conceal *q*. *e* should be predictable to the owner so that *v* is stable. Inspired by hardened child key derivation, the viewing key *v* could be deduced by the same algorithm as https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki?plain=1#L81

```js
const v = concat(q.chainCode, 00, q.privateKey, index)
       |> HMAC_SHA512
       |> slice(0, 64)
       |> sum(q.privateKey)
```

By doing so, the viewing key can be located by a specific key derivation path.

**Comment by quake** [comment]
If we use bip32 to generate v, we need to backup 32 bytes chain code when exporting and importing wallet. I think this is the same as backup one extra v, don't see any difference.

**Comment by Keith-CY** [comment]
> If we use bip32 to generate v, we need to backup 32 bytes chain code when exporting and importing wallet. I think this is the same as backup one extra v, don't see any difference.

Maybe this case is too specific to Neuron, or HD wallet, because chain code is included in the exported keystore, so the stealth meta-address can be derived again when wallet is imported.

**Comment by Keith-CY** [comment]
> # Neuron Product Planning
> ## Design Principles
> 1. Ease of use: Considering the advanced needs of stealth address as a user's asset management, making it an advanced function under one account will be more convenient for users to manage their assets while facilitating subsequent expansion.
> 2. Multiplicity: the need exists for users to manage different assets with multiple stealth addresses, thus a single wallet account can create multiple privacy addresses.
> 3. Help users avoid errors: need to consider user account and device change scenarios, need to add guidance when users add new stealth address accounts, export wallets.
> 
> ## Detailed planning
> 1. **Send page** - Add stealth transaction mode
> 
> Send page to add stealth transaction mode, the user needs to enter the stealth meta-address of the receiving asset and synthesize the address of the receiving asset according to the random number automatically generated by the software, and enter the asset to send.

I think the **stealth recipient-address** can be revealed on confirmation instead of being transformed on the send page so the **stealth meta-address** can be double-checked. But it can be labeled when **stealth recipient-address** is recognized.

> 2. **Add Stealth Address Account** - Users can create a Stealth Address account and receive and manage their assets under this page.
> 
> * Create stealth address accounts : Users can create 1-N stealth address accounts, and the software automatically generates corresponding secret keys and stealth meta-addresses.
> * Receiving stealth address assets : Users can scan for stealth transactions and get feedback by clicking on the receiving asset in their stealth address account.
> * Managing stealth assets : Users can view the assets under the account in the stealth account and unlock and send the assets.

Besides, balances of **stealth recipient-address**s can be aggregated and displayed as the balance of **stealth meta-address* in the address book.

**Comment by quake** [comment]
> Maybe this case is too specific to Neuron, or HD wallet, because chain code is included in the exported keystore, so the stealth meta-address can be derived again when wallet is imported.

I see, no more issues now

**Comment by homura** [comment]
[Samourai Wallet](https://samouraiwallet.com/paynym) is a wallet that supports [stealth addresses](https://github.com/Samourai-Wallet/samourai-wallet-android/blob/c71f21a631bbc69db2bd411f2905c7f141d1523f/Samourai-Wallet-features.md#stealth-addressingbip47-payment-channels) and BIP47 payment code, it may be a good inspiration for an HD wallet

**Comment by homura** [comment]
> compute the shared secret $c=H(vP)$

I am not sure I have misunderstood it. I suppose the _P_ might be a demon here when tons of _P_ are published to the public, and it might cause the scanner to be theoretically impossible to work because of too many possibilities, or we may downgrade to the way of a known address(_P_) book, and it's not a non-interactive way

**Comment by homura** [comment]
> I am not sure I have misunderstood it. I suppose the _P_ might be a demon here when tons of _P_ are published to the public, and it might cause the scanner to be theoretically impossible to work because of too many possibilities, or we may downgrade to the way of a known address(_P_) book, but it's not a non-interactive way

I might get it, the scanner should only scan the new stealth addresses, so it's fine

```ts
declare let newTx: Transaction;

for (let output of newTx.outputs) {
  if (!is_stealth(output)) return

  const {P, Q} = parse_stealth(output)
  const c = H(v, P)
  const derivedQ = publicKeyDerive(c+q)

  if(Q != derivedQ) return

  // continue to handle the logic as an own key
}

```

**Comment by Keith-CY** [comment]
> > compute the shared secret c=H(vP)
> 
> I am not sure I have misunderstood it. I suppose the _P_ might be a demon here when tons of _P_ are published to the public, and it might cause the scanner to be theoretically impossible to work because of too many possibilities, or we may downgrade to the way of a known address(_P_) book, but it's not a non-interactive way

I'm not sure what `demon` means. But `P` is generated by the sender as a nonce of the shared secret, and it's not for scanning. *V*, included in the **stealth meta-address**, is used for scanning. Infinite **stealth recipient-address**s can be derived from a single *V*, so it's unnecessary to generate infinite *V* for an account.

**Comment by Keith-CY** [comment]
> > I am not sure I have misunderstood it. I suppose the _P_ might be a demon here when tons of _P_ are published to the public, and it might cause the scanner to be theoretically impossible to work because of too many possibilities, or we may downgrade to the way of a known address(_P_) book, but it's not a non-interactive way
> 
> I might get it, the scanner should only scan the new stealth addresses, so it's fine
> 
> ```ts
> declare let newTx: Transaction
> declare let v: ViewingPrivateKey
> declare let q: SpendPrivateKey
> declare let H: HashFunction
> 
> for (let output of newTx.outputs) {
>   if (!is_stealth(output)) return
> 
>   const {P, Q} = parse_stealth(output)
>   const c = H(v, P)
>   const derived_Q = public_key_create(c+q)
> 
>   if(Q != derived_Q) return
> 
>   // continue to handle the logic as an own key
> }
> ```
> 

`if (Q != derived_Q) return` is infeasible here because `q` is private and cannot be used to check if an address is matched.

It should be

```js
for (let output of newTx.outputs) {
  if (!is_stealth(output)) return

  const { P, V' } = parse_stealth(output)
  const c = H(v.P)
  if ((V + c.G) !== V')) return // v is not matched
  const spendingPrivateKey = q + c
}
```


> But if we put the _P_ in the lock args, does it mean that the sender is transparent, and only the receiver is hidden?

*P* is an ephemeral nonce that reveals nothing. It's a part of the shared secret. But the sender can be located by `input` of the transaction.

**Comment by homura** [comment]
> I'm not sure what `demon` means. But _P_ is generated by the sender as a nonce of the shared secret, and it's not for scanning. _V_, included in the **stealth meta-address**, is used for scanning. Infinite **stealth recipient-address**s can be derived from a single _V_, so it's unnecessary to generate infinite _V_ for an account.

My first review skipped the _P_ in the recipient address, because I thought it was similar to a meta address which is without a _P_, like `Q' | V'`, since I mistook the _P_ for Alice's public key but it's just a random number. And I have no problem understanding that

**Comment by homura** [comment]
>   const spendingPrivateKey = q + c

It is important to check the _Q'_ in addition to the _V'_, as the _Q'_ represents the public key of the spending key. 
I would also like to confirm if `P | Q'` is enough. This is because by verifying if `publik_key_create(c + q) == Q'`, we can determine if the address is owned by the wallet.

**Comment by Keith-CY** [comment]
> > const spendingPrivateKey = q + c
> 
> It is important to check the _Q'_ in addition to the _V'_, as the _Q'_ represents the public key of the spending key. I would also like to confirm if `P | Q'` is enough. This is because by verifying if `publik_key_create(c + q) == Q'`, we can determine if the address is owned by the wallet.

Password would be required if `q` is used during scanning. It would be
1. Input password everytime an address to validate, which requires interaction; Or
2. Cache `q` in runtime, which is insecure.

So `v` is introduced to replace `q` during scanning.

**Comment by Danie0918** [comment]
【Support stealth address mode】

- Requirement Description: Supports stealth address mode.



![Image](https://github.com/Magickbase/neuron-public-issues/assets/122437870/285a9d92-96a8-4af3-8823-7e9c8584a429)




- Requirement Contents 
1. Add stealth address account entry to the Overview.
2. Add stealth address option switch to send page.
3. When sending a stealth transaction, you need to fill in the stealth receive meta address, and Neuron will generate the corresponding stealth receive address.
4. Added stealth address account page:
   - Automatically retrieve stealth address assets upon entry.
   - [Balance]: Show the total assets of all stealth address accounts.
   - [Update All]: Click to update all stealth address account assets.
   - [Stealth address account]: Show account name and meta-address.
   - [Copy]: Click to copy the meta-address.
   - [Switch]: Click to switch stealth address accounts.
   - [Stealth address accounts]: Supports renaming, switching, and deleting. （Only accounts with a balance of 0 can be deleted.）
5. Stealth address account asset list
   - Show the list of 'Receipt address ' 'Description' 'Balance' 'Txs' 'Operation' contents.
   - The default is to sort by the order of the last transaction in chronological order from latest to earliest.
   - Support 'Balance' and 'Txs' sorting.
   - 'Address' 'Balance' supports copy operations.
   - Supports adding/modifying 'Description'.
   - [Operation]: Support open explorer and sending.
   - [Batch operation]: Supports selecting addresses and sending them.

**Comment by yanguoyu** [comment]
I can not know the difference between `receive meta address` and `corresponding stealth receive address`. It may be unnecessary to show the `stealth receive address 1` that is generated from `stealth receive address `.
And I believe it's enough to generate only one stealth address for one wallet.

On the addtionly, it's better to consider how to optimate the sync of the stealth balance. Because we should scan all of the blocks one by one to recognize the stealth address balance.

**Comment by Danie0918** [comment]
> I can not know the difference between `receive meta address` and `corresponding stealth receive address`. It may be unnecessary to show the `stealth receive address 1` that is generated from `stealth receive address `. 

`receive meta address`: A stealth address provided by the receiver and not publicly available.
 `corresponding stealth receive address`: The actual receiving address based on the steganographic meta-address calculated by an algorithm, which is the address actually displayed in the transaction.

>And I believe it's enough to generate only one stealth address for one wallet.

The multiple stealth addresses are in consideration of the fact that users may need to categorize and manage assets.

> On the addtionly, it's better to consider how to optimate the sync of the stealth balance. Because we should scan all of the blocks one by one to recognize the stealth address balance.

Do you have any ideas about this? @Keith-CY

**Comment by yanguoyu** [comment]
Could you explain what these three marks mean to help me understand? @Danie0918

<img width="949" alt="image" src="https://github.com/Magickbase/neuron-public-issues/assets/12881040/05f5b55d-8c54-4f32-96a0-5ea0cc957135">

**Comment by Danie0918** [comment]
> Could you explain what these three marks mean to help me understand?

![new_page_1](https://github.com/Magickbase/neuron-public-issues/assets/122437870/29cd538e-d736-47f8-97ca-788dd91bdbfd)

1. Meta address name, the system default is Meta address + serial number.
2. The asset receiving address is attributed under the meta-address.
3. Account Addresses for the Stealth Address Account List

**Comment by yanguoyu** [comment]
I cannot understand the workflow of the `Stealth Address Account`.
1. How to create a `Receipt address`?
2. I guess this page will include many `Receipt address`. When does the `Meta Address 1 ` show at the top? (Select one Receipt Address?)
3. Will two `Meta Address` generate one `Receipt Address`?

**Comment by Danie0918** [comment]
> I cannot understand the workflow of the `Stealth Address Account`.
> 
> 1. How to create a `Receipt address`?
The `Receipt address` is obtained by scanning the chain for transactions and matching them, no creation is required.

> 2. I guess this page will include many `Receipt address`. When does the `Meta Address 1 ` show at the top? (Select one Receipt Address?)
Each meta-address can be treated as an account, and users can create multiple meta-addresses, each of which may have many assets in the `Receipt address`,

> 3. Will two `Meta Address` generate one `Receipt Address`?
As stated in 1, a `Receipt Address` is obtained by matching, and thus it will only attribute a meta-address.

**Comment by yanguoyu** [comment]
> > I cannot understand the workflow of the `Stealth Address Account`.
> > 
> > 1. How to create a `Receipt address`?
> 
> The `Receipt address` is obtained by scanning the chain for transactions and matching them, no creation is required.
> 
> > 2. I guess this page will include many `Receipt address`. When does the `Meta Address 1 ` show at the top? (Select one Receipt Address?)
> 
> Each meta-address can be treated as an account, and users can create multiple meta-addresses, each of which may have many assets in the `Receipt address`,
> 
> > 3. Will two `Meta Address` generate one `Receipt Address`?
> 
> As stated in 1, a `Receipt Address` is obtained by matching, and thus it will only attribute a meta-address.

Thanks, I got it. The `Meta Address` is the address that can be published to the network that can let everyone know. The `Receipt address` is the address that receives the assets that only the sender and receiver know.

**Comment by Keith-CY** [comment]
I think this workflow exposes too much about specific addresses, which is intended to be concealed by the account concept.

As a sender, only `stealth meta address` matters because
1. senders cannot validate the `stealth recipient address` with a glance;
2. the `stealth recipient address` varies because the generation includes a random number, `p` in the docs at https://docs.google.com/document/d/18CgG_bjweOx0ZgI-eFRFn5HYWOLWh-BFfjBQ-aRfLl8

As a receiver, all specific addresses only work for communication, telling others how to transfer to me. So the addresses should be dimmed on all pages. Why should I consider on which address the asset is held?

For me, the introduction of `stealth address` should be very simple

As a sender, when I fill in with a `stealth meta address`, it should be marked and annotated that the real address in the transaction will be different.

As a receiver, I can 
1. convert a general address into a `stealth meta address` irreversibly, and prevent users from copying the original address, e.g. show the `stealth meta address` instead of the `general address` in the address book;
2. view the balance of `stealth meta address`, which is the balance sum of its `stealth recipient address`;
3. mark the `stealth recipient address` in the transaction detail;
4. searching `stealth meta address` returns all transactions of its `stealth recipient address`
5. viewing `stealth recipient address` of a `stealth meta address` is a nice-to-have function, but should be inconspicous.

Mentioned the PRD above, following details are weird.

On the send page, there's a `stealth transaction` switch toggling the type of address to be inputted. So should all receiver addresses be stealth addresses or not? It's unnecessary.

On the stealth address account page, there's an `update all` button, which works as the same as `clear cache`

There's a `delete` button next to each meta address, which is not allowed. If the meta address is removed, 1. all assets will be undetectable; 2. the original address is potentially public as a general recipient address, make the used stealth address public;

Why are there `selected all` and `select some` on the `stealth meta address` page, there's no similar usage for the general address list.

**Comment by Keith-CY** [comment]
> > On the addtionly, it's better to consider how to optimate the sync of the stealth balance. Because we should scan all of the blocks one by one to recognize the stealth address balance.
> 
> Do you have any ideas about this? @Keith-CY

Stealth address cannot be detected in light client mode because it cannot be subscribed. So it's only supported in full node mode.

And the way to detect a stealth recipient address is a bit more consuming than the general address.

Check the section '扫描 stealth 交易的安全问题' at https://docs.google.com/document/d/18CgG_bjweOx0ZgI-eFRFn5HYWOLWh-BFfjBQ-aRfLl8/

`v` is introduced for scanning, named `viewing key`, when an output cell is under detection, `v`'s public key `V` will be used to check if it matches the tweaked `V'`. So computation will be made for each output cell. Thus all transactions should be scanned and only the full node works in this scenario.

**Comment by Danie0918** [comment]
![account_mode](https://github.com/Magickbase/neuron-public-issues/assets/122437870/454c9d75-994d-4f5d-8a10-bd79727eae50)

- Requirement Contents 
1. Add stealth address account entry to the Overview.
2. When sending, when the receiving address is recognized as a stealth meta-address, the transaction is automatically converted to a stealth transaction.
3. Added stealth address account page:
   - Automatically retrieve stealth address assets upon entry.
   - [Balance]: Show the total assets of all stealth address accounts.
   - [Update All]: Click to retrieve the stealth assets on the chain.
   - [Stealth address account]: Show account name and meta-address.
   - [Receive]: Click to open the receiving address. (Open creates one by default when there is no meta address.)
   - [Stealth address accounts]: Support for renaming and sorting. 
4. Stealth address account asset list
   - Show the list of 'Name ' 'Type' 'Amount''Balance' 'Time', and 'Operation' contents.
   - The default is to sort by the order of the last transaction in chronological order from latest to earliest.
   - Supports filtering list contents by 'Name'.
   - Support 'Balance' and 'Time' sorting.
   - 'Balance' supports copy operations.
   - Supports adding/modifying 'Description'.
   - [Operation]: Support open explorer and sending.
   - [Batch operation]: Supports selecting addresses and sending them.

**Comment by Keith-CY** [comment]
> - Requirement Contents 
> 1. Add stealth address account entry to the Overview.
> 2. When sending, when the receiving address is recognized as a stealth meta-address, the transaction is automatically converted to a stealth transaction.

This conversion could be done in the background instead of being displayed in UI because the real address derived from the stealth meta address means nothing to the sender.

> 3. Added stealth address account page:
>    - Automatically retrieve stealth address assets upon entry.
>    - [Balance]: Show the total assets of all stealth address accounts.

I found `stealth address account` was mentioned below and it's a meta-address, so the `total` here should be the sum of all `meta address`

>    - [Update All]: Click to retrieve the stealth assets on the chain.

Why not update all automatically but users have to update them manually

>    - [Stealth address account]: Show account name and meta-address.

Each address has its own `description` field (on the address list), maybe it can be used as the identity of `meta address`

>    - [Receive]: Click to open the receiving address. (Open creates one by default when there is no meta address.)

The usage of `stealth meta address` must be misunderstood. The `stealth meta address` is intended to be public, and the sender derives the concrete `stealth recipient address` based on the `stealth meta address`. So the generation happens on the sender side. What the receiver should reveal is the `stealth meta address`

>    - [Stealth address accounts]: Support for renaming and sorting. 
> 4. Stealth address account asset list
>    - Show the list of 'Name ' 'Type' 'Amount''Balance' 'Time', and 'Operation' contents.

`Amount` and `Balance` should be one field

>    - The default is to sort by the order of the last transaction in chronological order from latest to earliest.
>    - Supports filtering list contents by 'Name'.
>    - Support 'Balance' and 'Time' sorting.

Will non-token be excluded when sorted by `balance`?

>    - 'Balance' supports copy operations.
>    - Supports adding/modifying 'Description'.

We don't have `description` for each asset before, why it is introduced in the asset list of a stealth meta address

>    - [Operation]: Support open explorer and sending.
>    - [Batch operation]: Supports selecting addresses and sending them.
>

**Comment by Danie0918** [comment]
> This conversion could be done in the background instead of being displayed in UI because the real address derived from the stealth meta address means nothing to the sender.

Get it.

> I found `stealth address account` was mentioned below and it's a meta-address, so the `total` here should be the sum of all `meta address`

Yes.

> Why not update all automatically but users have to update them manually

The first time you enter it will be automatically updated, but to ensure real-time availability add the option for users to manually update it. 

> Each address has its own `description` field (on the address list), maybe it can be used as the identity of `meta address`

I don't get it.

> The usage of `stealth meta address` must be misunderstood. The `stealth meta address` is intended to be public, and the sender derives the concrete `stealth recipient address` based on the `stealth meta address`. So the generation happens on the sender side. What the receiver should reveal is the `stealth meta address`

Here it is being considered as a meta-address, I'll modify the description.

> `Amount` and `Balance` should be one field

Here 'Amount' is the number of assets and 'Balance' is equivalent to 'Capacity'

> Will non-token be excluded when sorted by `balance`?

Yes.

> We don't have `description` for each asset before, why it is introduced in the asset list of a stealth meta address

For the asset list, I'm biased toward equating it to 'History', showing transactions.

> What does `select addresses` mean when the action is performed to an asset list

The description here needs to be changed, it was meant to operate on assets.

**Comment by Keith-CY** [comment]
> The first time you enter it will be automatically updated, but to ensure real-time availability add the option for users to manually update it.

The balances should always be updated automatically

> > Each address has its own description field (on the address list), maybe it can be used as the identity of meta address
> I don't get it.

Now each `secp256k1 address` on the list has its own `description` field. When the `secp256k1 address` is turned into a `stealth meta address`, it will not be recovered, so its `description` field is useless because the address will never be used directly.

Then I see each `stealth meta address` needs an identity/account name, so I'm considering using the `description` field as its identity/account name.

To be specific, 
1. I pick an address as the `address meta address`
2. I update its description
3. description shown as its account name

> Here 'Amount' is the number of assets and 'Balance' is equivalent to 'Capacity'

It was mentioned in `Stealth address account asset list`. IMO, assets should include `CKB` and the list should be like

| name | type | amount | operation |
|-----|-|-|-|
|CKB| CKB | xxx | - |
|USDC | SUDT | xxx | - |

But you'd like to keep CKB counted separately as

| name | type | amount | balance | operation |
|-----|-|-|-|-|
|CKB| CKB | xxx | xxx | - |
|USDC | SUDT | xxx | yyy | - |

Am I right? But why separate CKB into various parts?

> For the asset list, I'm biased toward equating it to 'History', showing transactions.

Transactions related to `stealth recipient address` have their places in the "History", along with the `description` field, so why equate `asset list` to `history`

**Comment by Danie0918** [comment]
> The balances should always be updated automatically

Does this mean that this requires real-time querying of information on the chain?



> Now each `secp256k1 address` on the list has its own `description` field. When the `secp256k1 address` is turned into a `stealth meta address`, it will not be recovered, so its `description` field is useless because the address will never be used directly.
>
> Then I see each `stealth meta address` needs an identity/account name, so I'm considering using the `description` field as its identity/account name.
> 
> To be specific,
> 
> 1. I pick an address as the `address meta address`
> 2. I update its description
> 3. description shown as its account name

Get it.


> It was mentioned in `Stealth address account asset list`. IMO, assets should include `CKB` and the list should be like
> 
> name	type	amount	operation
> CKB	CKB	xxx	-
> USDC	SUDT	xxx	-
> But you'd like to keep CKB counted separately as
> 
> name	type	amount	balance	operation
> CKB	CKB	xxx	xxx	-
> USDC	SUDT	xxx	yyy	-
> Am I right? But why separate CKB into various parts?

Yes, 'Balance' is to show the breakdown field of the total balance. It is true that they are the same when the type is CKB, or we can consider providing an option to the user whether to show the balance field or not.
 
> Transactions related to `stealth recipient address` have their places in the "History", along with the `description` field, so why equate `asset list` to `history`

The list of assets is a chronological list of unspent stealth transactions with similarities to those recorded in history.

**Comment by Keith-CY** [comment]
> Does this mean that this requires real-time querying of information on the chain?

Stealth recipient address works as other addresses, and their balances keeps updating along with synchronization.

There's one exception. If the `stealth meta address` was declared otherwhere, and it's declared again in Neuron, Neuron has to re-sync the chain thoroughly to find out all live cells controlled by the `stealth meta address`. It can be avoided in terms of UX.

**Comment by Danie0918** [comment]
> <img alt="image" width="357" src="https://private-user-images.githubusercontent.com/12881040/297296988-1fca488d-8b09-403e-b64f-503f33f352fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU1NDUzNTEsIm5iZiI6MTcwNTU0NTA1MSwicGF0aCI6Ii8xMjg4MTA0MC8yOTcyOTY5ODgtMWZjYTQ4OGQtOGIwOS00MDNlLWI2NGYtNTAzZjMzZjM1MmZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE4VDAyMzA1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmYTE1NWIyMDE2OWE1MGY1YWRhYzIyZGQ3MmI3YjNlYTY4YWQ3YzFkMzMyNzMzMTQ0NWNjMjhhMzdiMzRhY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.N5H5sJ443LKIsljcuBkkgV6AtIFKJlEllNS4gboG39I">
> I guess for a `Meta Address`, there may include more than one transaction. Which transaction will show when expanding the meta address? @Danie0918

There is only 1 transaction here, so only the details of this transaction need to be shown.
